### PR TITLE
Refactor wrapper.ts to improve maintainability

### PR DESCRIPTION
**Pull request by stephensli**
# Why? 

This change works on refactoring out the wrapper.ts file to allow improved maintainability, extensibility with support of easy typing. The current implementation is one continuous function call which allocates all the bindings with hard to follow scoping.

Any bindings of methods have been split out into the `binding` directly, currently containing the core (anything not compiler related) and compiler related bound functions.

# Whats Next? 

Since the wrapper is the main entry point for most consumers of the service, this should allow cleanly defined types to be exported with the package. Allowing language servers and IDEs to consume the types and expose them to the user for an improved experience. 

* The callback implementation to apply changes based on some additional input arguments can now be tackled as it can be harder to follow. I was going to do it in this PR but it was outside of the scope. 

This has been done here #615

### CODE REVIEW

File: bindings/compile.ts [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+    return runWithCallbacks(solJson, core, readCallback, compileInternal, [input, optimize]);
+  };
+}
+
+/**
+ * Returns a binding to the solidity solidity_compile method with a fallback to
+ * compileStandard.
+ *
+ * @param solJson
+ * @param core
```
Comment (cameel) [inline bindings/compile.ts:60]: These should include parameter descriptions.

File: bindings/compile.ts [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+ */
+function compileStandard (solJson, core) {
```
Comment (cameel) [inline bindings/compile.ts:62]: Why do you really need this dependency injection here? I could see it for soljson since it's loaded at runtime but wouldn't it be simpler to just import `core`?

File: wrapper.ts [hunk 3]
```diff
@@ -1,360 +1,159 @@
...
-  let compileJSONMulti = null;
-  if ('_compileJSONMulti' in soljson) {
-    // input (jsontext), optimize (bool) -> output (jsontext)
-    compileJSONMulti = soljson.cwrap('compileJSONMulti', 'string', ['string', 'number']);
-  }
+  let input: { language: string, sources: any[], settings: any };
 
-  let compileJSONCallback = null;
-  if ('_compileJSONCallback' in soljson) {
-    // input (jsontext), optimize (bool), callback (ptr) -> output (jsontext)
```
Comment (cameel) [inline wrapper.ts:183]: The extra info from these comments (i.e. that the input and output is meant to be JSON) seems lost, even in the typed version.

File: bindings/core.ts [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+    // @ts-ignore
+    versionToSemver: versionToSemver(core.version())
+  };
+
+  return {
+    ...core,
+    ...helpers,
+
+    isVersion6: isVersion6(helpers.versionToSemver())
```
[the remark this answers is not in the document]
Comment (cameel) [reply bindings/core.ts:27]: Also, I just realized that this is not strictly version 6 but 6+. We should give it a more specific name:
```suggestion
    isVersion6OrNewer: isVersion6(helpers.versionToSemver())
```

**Review by cameel — COMMENTED** [review]
Unfortunately I did not manage to the through it all today so here's what I have so far.

Overall I did not find any real problems, mostly style and places where naming could be clearer.

File: bindings/compile.ts [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+/**********************
+ * COMPILE
+ **********************/
+
+/**
+ * Returns a binding to the solidity compileJSON method.
+ *
+ * @param solJson
+ */
+function compileJson (solJson) {
```
[the remark this answers is not in the document]
Comment (stephensli) [reply bindings/compile.ts:24]: Yeap, I agree and these have been updated to include a prefix of "bind"

File: bindings/compile.ts [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+/**
+ * Returns a binding to the solidity compileJSONCallback method.
+ *
+ * @param solJson
+ */
+function compileJsonCallback (solJson, core) {
+  const compileInternal = bindSolcMethod(solJson, 'compileJSONCallback',
+    'string', ['string', 'number', 'number'], null);
+
+  if (isNil(compileInternal)) return compileInternal;
```
[the remark this answers is not in the document]
Comment (stephensli) [reply bindings/compile.ts:48]: done

File: bindings/compile.ts (hunk 2 above)
Comment (stephensli) [reply bindings/compile.ts:62]: Not sure what you mean, core in this situation is used for the runWithCallbacks methods which in turn is used for reset, addFunction, removeFunction, isVersion6, etc.

File: bindings/compile.ts [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+  const compileInternal = bindSolcMethod(solJson, 'compileStandard',
+    'string', ['string', 'number'], null);
+
+  if (core.isVersion6) {
+    boundFunctionSolidity = bindSolcMethod(solJson, 'solidity_compile',
+      'string', ['string', 'number', 'number'], null);
+  } else {
+    boundFunctionSolidity = bindSolcMethod(solJson, 'solidity_compile',
+      'string', ['string', 'number'], null);
+  }
```
[the remark this answers is not in the document]
Comment (stephensli) [reply bindings/compile.ts:75]: Not my style but can do.

File: bindings/core.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (stephensli) [reply bindings/core.ts:27]: Yeah I saw this but I did not explicitly make the change, I will make it though

[the remark this answers is not in the document]
Comment (stephensli) [reply bindings/core.ts:27]: done

File: bindings/helpers.ts [hunk 8]
```diff
@@ -0,0 +1,39 @@
+import * as semver from 'semver';
+
+import { isNil } from '../common/helpers';
+
+export function bindSolcMethod (solJson, method, returnType, args, defaultValue) {
+  if (isNil(solJson[`_${method}`])) return defaultValue;
```
[the remark this answers is not in the document]
Comment (stephensli) [reply bindings/helpers.ts:6]: done

File: bindings/helpers.ts [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+  return solJson.cwrap(method, returnType, args);
+}
+
+export function bindSolcMethodWithFallbackFunc (solJson, method, returnType, args, fallbackMethod, finalFallback = null) {
+  const methodFunc = bindSolcMethod(solJson, method, returnType, args, null);
+  return methodFunc || bindSolcMethod(solJson, fallbackMethod, returnType, args, finalFallback);
```
[the remark this answers is not in the document]
Comment (stephensli) [reply bindings/helpers.ts:12]: Updated to perform explicit null check

File: bindings/helpers.ts [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+}
+
+export function getSupportedMethods (solJson) {
+  const anyMethods = anyMethodsExists.bind(this, solJson);
+
+  return {
+    license: anyMethods('solidity_license'),
```
[the remark this answers is not in the document]
Comment (stephensli) [reply bindings/helpers.ts:19]: It was just cleaner in my eyes, I have reverted it back

File: bindings/helpers.ts [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+    alloc: anyMethods('solidity_alloc'),
+    reset: anyMethods('solidity_reset'),
+    compileJson: anyMethods('compileJSON'),
+    compileJsonMulti: anyMethods('compileJSONMulti'),
+    compileJsonCallback: anyMethods('compileJSONCallback'),
+    compileJsonStandard: anyMethods('compileStandard', 'solidity_compile')
+  };
+}
+
+function anyMethodsExists (solJson, ...names) {
```
[the remark this answers is not in the document]
Comment (stephensli) [reply bindings/helpers.ts:30]: done

File: bindings/index.ts [hunk 12]
```diff
@@ -0,0 +1,15 @@
+import { setup as setupCore } from './core';
+import { getSupportedMethods } from './helpers';
+import { setup as setupCompile } from './compile';
+
+export default function setup (solJson) {
```
[the remark this answers is not in the document]
Comment (stephensli) [reply bindings/index.ts:5]: Yeah I agree, if you don't work with it a lot it can become unclear that you are looking for the default exported function in this situation.

[the remark this answers is not in the document]
Comment (stephensli) [reply bindings/index.ts:5]: done

File: tsconfig.json [hunk 13]
```diff
@@ -10,19 +10,20 @@
...
+
     // Allow JS must be included to ensure that the built binary is included
     // in the output. This could be copied directly in the future if required.
     "allowJs": true,
     // TODO:
     // In order to gracefully move our project to TypeScript without having
     // TS immediately yell at you, we'll disable strict mode for now.
     "strict": false,
-    "noImplicitAny": false
+    "noImplicitAny": false,
```
[the remark this answers is not in the document]
Comment (stephensli) [reply tsconfig.json:21]: This mainly exists from a removal of required changes for the type export which is now a separate PR, i have now resolved this.

[the remark this answers is not in the document]
Comment (stephensli) [reply tsconfig.json:21]: done

File: bindings/compile.ts (hunk 1 above)
Comment (stephensli) [reply bindings/compile.ts:76]: They could but they would have self describing types in the following PR so I did not see it as a requirement. I can add the additional comments if you feel they are required here.

File: wrapper.ts [hunk 14]
```diff
@@ -1,360 +1,159 @@
...
+      compileSingle: compile.compileJson,
+      compileMulti: compile.compileJsonMulti,
+      compileCallback: compile.compileJsonCallback,
+      compileStandard: compile.compileStandard
+    },
+    features: {
+      legacySingleInput: supported.compileJson,
+      multipleInputs: supported.compileJsonMulti || supported.compileJsonStandard,
+      importCallback: supported.compileJsonCallback || supported.compileJsonStandard,
+      nativeStandardJSON: supported.compileJsonStandard
```
[the remark this answers is not in the document]
Comment (stephensli) [reply wrapper.ts]: Added a suffix of "Supported"

File: wrapper.ts [hunk 15]
```diff
@@ -1,360 +1,159 @@
...
+function loadRemoteVersion (versionString, callback) {
+  const memoryStream = new MemoryStream(null, { readable: false });
+  const url = `https://binaries.soliditylang.org/bin/soljson-${versionString}.js`;
 
-  // This calls compile() with args || cb
-  const runWithCallbacks = function (callbacks, compile, args) {
-    if (callbacks) {
-      assert(typeof callbacks === 'object', 'Invalid callback object specified.');
+  https.get(url, response => {
+    if (response.statusCode > 299 || response.statusCode < 200) {
```
[the remark this answers is not in the document]
Comment (stephensli) [reply wrapper.ts:44]: Ah the http codes, yeah I can convert it and add it again in another pr in the future.

File: wrapper.ts (hunk 3 above)
Comment (stephensli) [reply wrapper.ts:183]: I did not feel them a need since they would be explicitly typed in the following PR but I will add them back anyway.

Comment (stephensli) [reply wrapper.ts:183]: done

File: wrapper.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (cameel) [reply wrapper.ts]: I mean, a separate commit is enough. Does not have to be a full-blown PR. It's just stuff that's easy to miss in a bigger refactor :)

File: wrapper.ts (hunk 3 above)
Comment (cameel) [reply wrapper.ts:183]: I mean, there isn't a type for JSON is there? It's just a string. The rest is there but the fact that input and output JSON should be kept somewhere in the docstring.

File: bindings/compile.ts (hunk 2 above)
Comment (cameel) [reply bindings/compile.ts]: I mean that you could simply import `core` at the top of `compiler.ts` and not have to pass it as an argument. This is a valid design, especially when you need to isolate things in tests but it also makes the code less straightforward so I'm just wondering if it's worth it here. Is there a particular reason you chose that the design here?

Comment (cameel) [reply bindings/compile.ts]: Ah, I see what you meant. Looking through it again now I realized that the `core` parameter here is not the `core.ts` module as I assumed.

How about naming these parameters/variables `coreBindings`, `compileBindings` and `supportedMethods`? In the PR we can see the type which makes it better but I think that the name should be readable on its own too.

File: wrapper.ts [hunk 16]
```diff
@@ -1,360 +1,159 @@
...
+    return formatFatalError('No input sources specified.');
+  }
 
-    try {
-      input = JSON.parse(input);
-    } catch (e) {
-      return formatFatalError('Invalid JSON supplied: ' + e.message);
-    }
+  const sources = translateSources(input);
+  const optimized = isOptimizerEnabled(input);
```
Comment (cameel) [inline wrapper.ts:91]:
```suggestion
  const optimize = isOptimizerEnabled(input);
```

File: wrapper.ts [hunk 17]
```diff
@@ -1,360 +1,159 @@
...
-            soljson._compile(mem.toString(), 'soljson-' + versionString + '.js');
-            if (module.parent && module.parent.children) {
-              // Make sure the module is plugged into the hierarchy correctly to have parent
-              // properly garbage collected.
-              module.parent.children.splice(module.parent.children.indexOf(soljson), 1);
-            }
+function translateOutput (outputRaw, libraries) {
+  try {
+    const parsedOutput = JSON.parse(outputRaw);
+    const output = translate.translateJsonCompilerOutput(parsedOutput, libraries);
```
Comment (cameel) [inline wrapper.ts:150]: This `try` block is now too broad. It's only supposed to cover `JSON.parse()`. An exception on one of the other lines does not necessarily mean that invalid JSON was returned.

File: bindings/index.ts [hunk 18]
```diff
@@ -0,0 +1,15 @@
+import { setupCore } from './core';
+import { getSupportedMethods } from './helpers';
+import { setupCompile } from './compile';
+
+export default function setupBindings (solJson) {
+  const core = setupCore(solJson);
+  const compile = setupCompile(solJson, core);
+  const supported = getSupportedMethods(solJson);
```
Comment (cameel) [inline bindings/index.ts:8]: This might also be a good name:
```suggestion
  const methodFlags = getSupportedMethods(solJson);
```

File: bindings/core.ts [hunk 19]
```diff
@@ -0,0 +1,130 @@
+import { bindSolcMethod, bindSolcMethodWithFallbackFunc } from './helpers';
+import translate from '../translate';
+import * as semver from 'semver';
+
+export function setupCore (solJson) {
+  const core = {
+    alloc: bindAlloc(solJson),
+    license: bindLicense(solJson),
+    version: bindVersion(solJson),
+    reset: bindReset(solJson)
```
Comment (cameel) [inline bindings/core.ts:10]: Here are some thoughts on the overall structure:
- I don't think `license` and `version` really fit here. These are exposed by the wrapper while all the other methods in `core` are just helpers used by other methods. I'd move them to a separate module (could be e.g. `compiler-info.ts` or maybe `identify.ts`?).
- I'd rename `core.ts` to `internal.ts` because otherwise it's a bit weird that compilation is not in core - that's core functionality of a compiler.
- Why not put callbacks in a separate module?

**Review by cameel — COMMENTED** [review]
I finally managed to get through it all.

I found two or three places that look like bugs. This indicates that we're really don't have enough test coverage here.

Other than that, some comments about naming and overall structure. Some details need tweaking but I really like the final effect.

File: bindings/compile.ts (hunk 1 above)
Comment (cameel) [reply bindings/compile.ts:79]: I just think that these are pointless if they do not add any extra info on top of name and type. We already have name in the signature and we'll have type as well. If we add them, we should at least try to make them informative.

These two maybe are not best examples because they have unique types (though `SolJson` interface does not have a general description so saying that it's the compiler binary is still useful). But `bindCopyToCString()` for example does not have descriptions for parameters either even though `str` and `ptr` are not very informative on their own.

File: wrapper.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (stephensli) [reply wrapper.ts]: okay i'll add another commit at the end of this review to include it

File: bindings/compile.ts (hunk 2 above)
Comment (stephensli) [reply bindings/compile.ts]: yeah sounds good.

File: wrapper.ts (hunk 3 above)
Comment (stephensli) [reply wrapper.ts:183]: yeah I have added them

File: bindings/core.ts [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+
+/**
+ * Returns a binding to the solidity_license method.
+ *
+ * @param solJson
+ */
+function bindLicense (solJson) {
+  return bindSolcMethodWithFallbackFunc(solJson, 'solidity_license', 'string',
+    ['number'], 'license', () => {
+    });
```
[the remark this answers is not in the document]
Comment (stephensli) [reply bindings/core.ts:68]: Yeah I must have missed this one, I will update it.

File: bindings/core.ts [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+
+/**
+ * Returns a binding to the solidity_version method.
+ *
+ * @param solJson
+ */
+function bindVersion (solJson) {
+  return bindSolcMethodWithFallbackFunc(solJson, 'solidity_version', 'string',
+    ['number'], 'version', null);
+}
```
[the remark this answers is not in the document]
Comment (stephensli) [reply bindings/core.ts:54]: Yeah I can validate, this was just here to be more explicit with what is being returned on callback. Instead of just having to know the implementation detail of cwrap.

File: bindings/core.ts (hunk 20 above)
[the remark this answers is not in the document]
Comment (stephensli) [reply bindings/core.ts:68]: done

File: bindings/core.ts [hunk 22]
```diff
@@ -0,0 +1,130 @@
...
+
+/**
+ * Returns a binding to the solidity_license method.
+ *
+ * @param solJson
+ */
+function bindLicense (solJson) {
+  return bindSolcMethodWithFallbackFunc(solJson, 'solidity_license', 'string',
+    ['number'], 'license', () => {
```
[the remark this answers is not in the document]
Comment (stephensli) [reply bindings/core.ts:67]: done

[the remark this answers is not in the document]
Comment (stephensli) [reply bindings/core.ts:67]: done

File: bindings/core.ts [hunk 23]
```diff
@@ -0,0 +1,130 @@
...
+
+/**
+ * Returns a binding to the solidity_alloc function.
+ *
+ * @param solJson
+ */
+function bindAlloc (solJson) {
+  return bindSolcMethodWithFallbackFunc(solJson,
+    'solidity_alloc', 'number', ['number'], 'malloc');
+}
```
[the remark this answers is not in the document]
Comment (stephensli) [reply bindings/core.ts:44]: I'm not sure what you are trying to say here, this is falling back to malloc if solidity_alloc is not present.

File: wrapper.ts (hunk 16 above)
Comment (stephensli) [reply wrapper.ts:91]: done

File: wrapper.ts (hunk 17 above)
Comment (stephensli) [reply wrapper.ts:150]: sure I can simplify it.

File: bindings/core.ts (hunk 19 above)
Comment (stephensli) [reply bindings/core.ts:10]: Can do but callbacks are not bindings they are directly related to the compiling and should stay related (imo), will make this kind of change.

Comment (stephensli) [reply bindings/core.ts:10]: @cameel this creates a lot of additional implementation detail that adds more complexity than required. Taking those out results in passing another set of values deep into the callback code which is already complex.

File: bindings/core.ts [hunk 24]
```diff
@@ -0,0 +1,130 @@
...
+/**
+ * Copy from C String.
+ *
+ * is to support multiple versions of Emscripten.
+ * Take a single `ptr` and returns a `str`.
+ *
+ * @param solJson
+ * @param data
+ */
+function bindCopyFromCString (solJson, data) {
```
[the remark this answers is not in the document]
Comment (stephensli) [reply bindings/core.ts:144]: done

File: bindings/core.ts [hunk 25]
```diff
@@ -0,0 +1,130 @@
...
+  const buffer = alloc(length + 1);
+
+  solJson.stringToUTF8(str, buffer, length + 1);
+  solJson.setValue(ptr, buffer, '*');
+}
+
+/**
+ * Copy from C String.
+ *
+ * is to support multiple versions of Emscripten.
```
[the remark this answers is not in the document]
Comment (stephensli) [reply bindings/core.ts:138]: done

File: bindings/core.ts [hunk 26]
```diff
@@ -0,0 +1,130 @@
...
+}
+
+/**********************
+ * Helpers Functions
+ **********************/
+
+/**
+ * Copy to a C string.
+ *
+ * is allocating memory using solc's allocator.
```
[the remark this answers is not in the document]
Comment (stephensli) [reply bindings/core.ts:112]: done

File: bindings/core.ts [hunk 27]
```diff
@@ -0,0 +1,130 @@
...
+    version: bindVersion(solJson),
+    reset: bindReset(solJson)
+  };
+
+  const helpers = {
+    addFunction: bindAddFunction.bind(this, solJson),
+    removeFunction: bindRemoveFunction.bind(this, solJson),
+
+    copyFromCString: bindCopyFromCString.bind(this, solJson),
+    copyToCString: bindCopyToCString.bind(this, solJson, core.alloc),
```
[the remark this answers is not in the document]
Comment (stephensli) [reply bindings/core.ts:18]: done

File: bindings/core.ts (hunk 19 above)
Comment (cameel) [reply bindings/core.ts:10]: ok. I thought moving them would be easy but if it does introduce that much complexity then let's leave them as is.

I'll take a closer look when I'm back next week but overall this is not a deal breaker.

File: bindings/core.ts (hunk 23 above)
[the remark this answers is not in the document]
Comment (stephensli) [reply bindings/core.ts]: don

**Review by matheusaaguiar — APPROVED** [review]
I have checked the old code against the new one to ensure that every function is present and it seems fine.

File: bindings/core.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (cameel) [reply bindings/core.ts]: Have you checked this? Just to be clear - I'm not sure what `cwrap()` will actually do. I'm just saying that *if* if returns `null` but I don't know if it does.

File: bindings/core.ts (hunk 19 above)
Comment (cameel) [reply bindings/core.ts:11]: So apart from callbacks there are still the first 2 points. `core` really does not sound right to me here since the stuff in it is so peripheral. Maybe it would make more sense as `memory` (containing alloc, reset, the string helpers) and `info` (containing license, version, semver, etc.). Or something like that.

The only way it makes sense to me to call these `core` is by saying they're all equal parts of [`libsolc.h`](https://github.com/ethereum/solidity/blob/develop/libsolc/libsolc.h), which is true but still...

But anyway, at this point I just want to get this merged so that's just an optional suggestion. If you can adjust it easily, great, if not, we'll go with what it's already there anyway.

File: wrapper.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (cameel) [reply wrapper.ts]: Marking this unresolved since the commit is not there yet.

**Review by cameel — APPROVED** [review]
I left 3 comments but everything important has been addressed. I'm marking as approved since the review is done and we just need to finish discussing these small points and merge.

**Comment by cameel** [comment]
Also, please rebase on latest `master`. My test PRs went in recently.

File: bindings/core.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (stephensli) [reply bindings/core.ts]: I believe its been updated to return correctly, its been a while since I have had a look but I will double check it.

File: wrapper.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (stephensli) [reply wrapper.ts]: Ah yeah can do or another PR to keep it out, I will look to add this once we are fully happy with the other commits. I'm not a massive fan of this minimise commits flow when you can just squash commits during the merge which is why I left it out.

**Comment by r0qs** [comment]
Hey, what is missing in this PR to be merged? Just the rebase on the latest master? It would be great to have this merged.

**Comment by cameel** [comment]
Basically rechecking the 3 comments I mentioned and rebasing. We're pretty much done here.

**Comment by cameel** [comment]
Then we need to review #615.

**Comment by stephensli** [comment]
@r0qs and @cameel sorry guys, I have been super busy at work but its now on my plan to complete this on the weekend!

File: bindings/core.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (stephensli) [reply bindings/core.ts]: @cameel this has been updated to only handle fallback values which are not undefined. `bindVersion` now does not pass a final fallback which in turn forces the cwrap call to keep in line with the current implementation.

File: bindings/core.ts (hunk 19 above)
Comment (stephensli) [reply bindings/core.ts:11]: I believe this still stands.

**Review by cameel — APPROVED** [review]
Thanks! I think we can merge it now and move on to the PR that adds types.

**Comment by cameel** [comment]
Merged.
